File: app/src/main/java/com/kira/kdownloader/engine/FormatSelector.java
package com.kira.kdownloader.engine;

import java.net.URI;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

public final class FormatSelector {
    public static final String FALLBACK_VIDEO_SELECTOR = "bestvideo*+bestaudio/best";
    private static final List<Integer> PREFERRED_HEIGHTS = Arrays.asList(1080, 720, 480, 360);
    private static final Set<String> SINGLE_QUALITY_HOSTS = new HashSet<>(Arrays.asList(
            "instagram.com", "instagr.am", "tiktok.com", "facebook.com", "fb.com", "fb.watch"));

    public enum Kind { VIDEO, AUDIO }

    private FormatSelector() {}

    public static List<DownloadChoice> choices(List<FormatInput> formats) { return choices(formats, ""); }

    public static List<DownloadChoice> choices(List<FormatInput> formats, String sourceUrl) {
        if (usesSingleVideoQuality(sourceUrl)) {
            return singleQualityChoices(formats, ExtractorOptions.isTikTokUrl(sourceUrl));
        }
        Set<Integer> heights = new HashSet<>();
        for (FormatInput format : formats) if (hasVideo(format) && format.getHeight() != null) heights.add(format.getHeight());
        Long audioBytes = bestAudioBytes(formats);
        List<DownloadChoice> result = new ArrayList<>();
        for (Integer height : PREFERRED_HEIGHTS) {
            if (!heights.contains(height)) continue;
            result.add(new DownloadChoice(
                    height + "p", Kind.VIDEO,
                    "bestvideo[height<=" + height + "]+bestaudio/best[height<=" + height + "]",
                    null, Collections.emptyMap(), videoBytes(formats, height, audioBytes)));
        }
        result.add(audioChoice(audioBytes));
        return result;
    }

    static boolean usesSingleVideoQuality(String sourceUrl) {
        try {
            String host = new URI(sourceUrl.trim()).getHost();
            if (host == null) return false;
            host = host.toLowerCase(Locale.ROOT);
            for (String root : SINGLE_QUALITY_HOSTS) {
                if (host.equals(root) || host.endsWith("." + root)) return true;
            }
        } catch (Exception ignored) {
            return false;
        }
        return false;
    }

    private static Long bestAudioBytes(List<FormatInput> formats) {
        Long best = null;
        for (FormatInput format : formats) {
            Long size = format.getFilesize();
            if (hasAudio(format) && !hasVideo(format) && size != null && (best == null || size > best)) best = size;
        }
        return best;
    }

    private static Long videoBytes(List<FormatInput> formats, int height, Long audioBytes) {
        Long videoBytes = null;
        boolean alreadyMuxed = false;
        for (FormatInput format : formats) {
            if (!hasVideo(format) || format.getHeight() == null || format.getHeight() != height) continue;
            alreadyMuxed |= hasAudio(format);
            Long size = format.getFilesize();
            if (size != null && (videoBytes == null || size > videoBytes)) videoBytes = size;
        }
        if (videoBytes == null || alreadyMuxed) return videoBytes;
        return videoBytes + (audioBytes == null ? 0 : audioBytes);
    }

    private static List<DownloadChoice> singleQualityChoices(List<FormatInput> formats, boolean allowDirectVideo) {
        Long audioBytes = bestAudioBytes(formats);
        boolean preferDirectCombined = false;
        for (FormatInput format : formats) {
            if (allowDirectVideo && hasVideo(format) && hasAudio(format) && notBlank(format.getUrl())) {
                preferDirectCombined = true;
                break;
            }
        }

        FormatInput best = null;
        for (FormatInput candidate : formats) {
            if (!hasVideo(candidate)) continue;
            if (best == null || compareVideo(candidate, best, preferDirectCombined) >= 0) best = candidate;
        }

        List<DownloadChoice> result = new ArrayList<>();
        if (best != null) {
            String selector = FALLBACK_VIDEO_SELECTOR;
            if (notBlank(best.getFormatId()) && allowDirectVideo) {
                selector = hasAudio(best) ? best.getFormatId() : best.getFormatId() + "+bestaudio/best";
            }
            String directUrl = allowDirectVideo && hasAudio(best) && notBlank(best.getUrl()) ? best.getUrl() : null;
            Long size = best.getFilesize();
            Long approxBytes = size == null ? null : size + (hasAudio(best) || audioBytes == null ? 0 : audioBytes);
            result.add(new DownloadChoice(
                    best.getHeight() == null ? "Video" : best.getHeight() + "p",
                    Kind.VIDEO, selector, directUrl,
                    directUrl == null ? Collections.emptyMap() : best.getHttpHeaders(), approxBytes));
        }
        result.add(audioChoice(audioBytes));
        return result;
    }

    private static int compareVideo(FormatInput left, FormatInput right, boolean preferDirectCombined) {
        if (preferDirectCombined) {
            int direct = Boolean.compare(hasAudio(left) && notBlank(left.getUrl()), hasAudio(right) && notBlank(right.getUrl()));
            if (direct != 0) return direct;
        }
        int height = Integer.compare(left.getHeight() == null ? 0 : left.getHeight(), right.getHeight() == null ? 0 : right.getHeight());
        if (height != 0) return height;
        return preferDirectCombined ? 0 : Boolean.compare(hasAudio(left), hasAudio(right));
    }

    private static DownloadChoice audioChoice(Long approxBytes) {
        return new DownloadChoice("Audio (mp3)", Kind.AUDIO, "bestaudio/best", null, Collections.emptyMap(), approxBytes);
    }

    private static boolean hasVideo(FormatInput format) {
        return format.getVcodec() != null && !"none".equals(format.getVcodec());
    }

    private static boolean hasAudio(FormatInput format) {
        return format.getAcodec() != null && !"none".equals(format.getAcodec());
    }

    private static boolean notBlank(String value) { return value != null && !value.trim().isEmpty(); }
}

File: app/src/main/java/com/kira/kdownloader/engine/FormatSelector.kt
package com.kira.kdownloader.engine

import java.net.URI

data class FormatInput(
    val formatId: String,
    val ext: String,
    val height: Int?,
    val vcodec: String?,
    val acodec: String?,
    val url: String? = null,
    val httpHeaders: Map<String, String> = emptyMap(),
    /** Reported (or approximate) size of this individual format in bytes, if known. */
    val filesize: Long? = null,
)

data class DownloadChoice(
    val label: String,
    val kind: FormatSelector.Kind,
    val formatSelector: String,
    val directUrl: String? = null,
    val httpHeaders: Map<String, String> = emptyMap(),
    /** Best-effort estimated download size in bytes for display, if known. */
    val approxBytes: Long? = null,
)

object FormatSelector {
    enum class Kind { VIDEO, AUDIO }

    private val preferredHeights = listOf(1080, 720, 480, 360)
    private val singleQualityHosts = setOf(
        "instagram.com",
        "instagr.am",
        "tiktok.com",
        "facebook.com",
        "fb.com",
        "fb.watch",
    )

    fun choices(formats: List<FormatInput>, sourceUrl: String = ""): List<DownloadChoice> {
        if (usesSingleVideoQuality(sourceUrl)) {
            return singleQualityChoices(
                formats = formats,
                allowDirectVideo = ExtractorOptions.isTikTokUrl(sourceUrl),
            )
        }

        val videoHeights = formats
            .asSequence()
            .filter { format -> format.hasVideo() }
            .mapNotNull { it.height }
            .toSet()

        val audioEstimate = bestAudioBytes(formats)
        val videoChoices = preferredHeights
            .filter(videoHeights::contains)
            .map { height ->
                DownloadChoice(
                    label = "${height}p",
                    kind = Kind.VIDEO,
                    formatSelector =
                        "bestvideo[height<=$height]+bestaudio/best[height<=$height]",
                    approxBytes = videoBytes(formats, height, audioEstimate),
                )
            }

        return videoChoices + audioChoice(audioEstimate)
    }

    /** Largest known audio-only size, used as the audio-track estimate for merged video. */
    private fun bestAudioBytes(formats: List<FormatInput>): Long? = formats
        .filter { it.hasAudio() && !it.hasVideo() }
        .mapNotNull(FormatInput::filesize)
        .maxOrNull()

    /** Estimated size for a video choice at [height]: the video format plus an audio track. */
    private fun videoBytes(
        formats: List<FormatInput>,
        height: Int,
        bestAudioBytes: Long?,
    ): Long? {
        val atHeight = formats.filter { it.hasVideo() && it.height == height }
        val videoBytes = atHeight.mapNotNull(FormatInput::filesize).maxOrNull() ?: return null
        val alreadyMuxed = atHeight.any { it.hasAudio() }
        return if (alreadyMuxed) videoBytes else videoBytes + (bestAudioBytes ?: 0)
    }

    private fun singleQualityChoices(
        formats: List<FormatInput>,
        allowDirectVideo: Boolean,
    ): List<DownloadChoice> {
        val audioEstimate = bestAudioBytes(formats)
        val preferDirectCombined = allowDirectVideo && formats.any { format ->
            format.hasVideo() && format.hasAudio() && !format.url.isNullOrBlank()
        }
        val bestVideo = formats
            .withIndex()
            .filter { (_, format) -> format.hasVideo() }
            .maxWithOrNull(
                if (preferDirectCombined) {
                    compareBy(
                        { (_, format) -> if (format.hasAudio() && !format.url.isNullOrBlank()) 1 else 0 },
                        { (_, format) -> format.height ?: 0 },
                        IndexedValue<FormatInput>::index,
                    )
                } else {
                    compareBy(
                        { (_, format) -> format.height ?: 0 },
                        { (_, format) -> if (format.hasAudio()) 1 else 0 },
                        IndexedValue<FormatInput>::index,
                    )
                },
            )
            ?.value

        val videoChoice = bestVideo?.let { format ->
            val selector = format.formatId
                .takeIf(String::isNotBlank)
                ?.let { formatId ->
                    when {
                        allowDirectVideo && format.hasAudio() -> formatId
                        allowDirectVideo -> "$formatId+bestaudio/best"
                        else -> FALLBACK_VIDEO_SELECTOR
                    }
                }
                ?: FALLBACK_VIDEO_SELECTOR

            val directUrl = format.url
                ?.takeIf { url -> allowDirectVideo && format.hasAudio() && url.isNotBlank() }
            val approxBytes = format.filesize?.let { vBytes ->
                if (format.hasAudio()) vBytes else vBytes + (audioEstimate ?: 0)
            }
            DownloadChoice(
                label = format.height?.let { "${it}p" } ?: "Video",
                kind = Kind.VIDEO,
                formatSelector = selector,
                directUrl = directUrl,
                httpHeaders = if (directUrl != null) format.httpHeaders else emptyMap(),
                approxBytes = approxBytes,
            )
        }

        return listOfNotNull(videoChoice) + audioChoice(audioEstimate)
    }

    internal fun usesSingleVideoQuality(sourceUrl: String): Boolean {
        val host = runCatching { URI(sourceUrl.trim()).host }
            .getOrNull()
            ?.lowercase()
            ?: return false

        return singleQualityHosts.any { root -> host == root || host.endsWith(".$root") }
    }

    private fun FormatInput.hasVideo(): Boolean =
        vcodec != null && vcodec != "none"

    private fun FormatInput.hasAudio(): Boolean =
        acodec != null && acodec != "none"

    private fun audioChoice(approxBytes: Long? = null) = DownloadChoice(
        label = "Audio (mp3)",
        kind = Kind.AUDIO,
        formatSelector = "bestaudio/best",
        approxBytes = approxBytes,
    )

    const val FALLBACK_VIDEO_SELECTOR = "bestvideo*+bestaudio/best"
}
